otter/convergence/gathering.py

```diff
@@ -41,30 +47,37 @@ def get_all_server_details(changes_since=None, batch_size=100):
     url = append_segments('servers', 'detail')
     query = {'limit': batch_size}
     if changes_since is not None:
-        query['changes_since'] = '{}Z'.format(changes_since.isoformat())
+        query['changes_since'] = '{0}Z'.format(changes_since.isoformat())
+    url = "{0}?{1}".format(url, urlencode(query))
+
+    last_link = []
 
-    def get_server_details(marker):
-        # sort based on query name to make the tests predictable
-        if marker is not None:
-            query.update({'marker': marker})
-        urlparams = sorted(query.items())
+    def get_server_details(url_with_query):
         eff = retry_effect(
-            service_request(
-                ServiceType.CLOUD_SERVERS,
-                'GET', '{}?{}'.format(url, urlencode(urlparams))),
+            service_request(ServiceType.CLOUD_SERVERS, 'GET',
+                            url_with_query),
             retry_times(5), exponential_backoff_interval(2))
         return eff.on(continue_)
 
     def continue_(result):
         _response, body = result
         servers = body['servers']
-        if len(servers) < batch_size:
-            return servers
-        else:
-            more_eff = get_server_details(servers[-1]['id'])
+        # Only continue if pagination is supported and there is another page
+        continuation = [link['href'] for link in body.get('servers_links', [])
+                        if link['rel'] == 'next']
+        if continuation:
+            # blow up if we try to fetch the same link twice
+            if last_link and last_link[-1] == continuation[0]:
+                raise UnexpectedBehaviorError(
+                    "When gathering server details, got the same 'next' link "
+                    "twice from Nova: {0}".format(last_link[-1]))
+
+            last_link.append(continuation[0])
```

Comment:
This could just be `last_link[:] = [continuation[0]]` to avoid keeping old links around, since we don't use them.